torch/_inductor/codegen/common.py

```diff
@@ -36,6 +44,114 @@ def index_prevent_reordering(index: typing.List[sympy.Expr], index_vars, sizes):
+            "logical_not",
+            "signbit",
+            "le",
+            "lt",
+            "ge",
+            "gt",
+            "eq",
+            "ne",
+            "bitwise_not",
+            "bitwise_or",
+            "bitwise_left_shift",
+            "bitwise_right_shift",
+        ]
+        ops_with_dtype_arg = ["constant", "to_dtype", "rand", "randn"]
+        reduction_to_bool = ["any"]
+        reduction_to_int64 = ["argmin", "argmax"]
+        ops_without_dtype = ["ops", "get_index"]
+        if OptimizationContext.key in _node.meta:
+            opt_ctx = _node.meta[OptimizationContext.key]
+        else:
+            opt_ctx = OptimizationContext()
+        if opt_ctx.dtype is not None:
+            return False
+        if _node.target in ops_to_bool:
+            opt_ctx.dtype = torch.bool
+        elif _node.target in ops_with_dtype_arg:
+            opt_ctx.dtype = _node.args[-1]
+        elif _node.target == "reduction":
+            reduction_type = _node.args[4]
+            if reduction_type in reduction_to_bool:
+                opt_ctx.dtype = torch.bool
+            elif reduction_type in reduction_to_int64:
+                opt_ctx.dtype = torch.int64
+        elif _node.target == "load":
+            opt_ctx.dtype = V.graph.get_dtype(_node.args[1])
+        if opt_ctx.dtype is not None:
+            data_type_logger(
+                f"for node.target = {_node.target}, dtype is propagated to {opt_ctx.dtype}"
+            )
+            _node.meta[OptimizationContext.key] = opt_ctx
+            return True
+
+        # node.target not belong to any ops which can directly get the dtype
+        # need propogate dtype with it's input node
+        dtype = None
+        inputs = node.all_input_nodes
+        input_nodes = [
+            n
+            for n in inputs
+            if isinstance(n, torch.fx.node.Node) and n.target not in ops_without_dtype
+        ]
+        if len(input_nodes) == 0:
+            return False
+        all_input_nodes_propogated = all(
+            OptimizationContext.key in n.meta
+            and n.meta[OptimizationContext.key].dtype is not None
+            for n in input_nodes
+        )
+        if not all_input_nodes_propogated:
+            return False
+        # all input nodes have propogated dtype, we will promot to dtype with highest precision
+        dtype = functools.reduce(
+            torch.promote_types,
+            [n.meta[OptimizationContext.key].dtype for n in input_nodes],
+        )
+        opt_ctx.dtype = dtype
+        msg = f"for node.target = {_node.target}, dtype is propagated to {opt_ctx.dtype}, "
+        input_msg = "inputs dtypes: "
+        for n in input_nodes:
+            input_msg += (
+                f"input {n.name}.dtype = {n.meta[OptimizationContext.key].dtype}"
+            )
+        data_type_logger(msg + input_msg)
+        _node.meta[OptimizationContext.key] = opt_ctx
+        opt_ctx.dtype_propogated = True
```

Comment:
Is this a duplicate of `opt_ctx.dtype`? Can we check just check if `opt_ctx.dtype` is `None` or not?

Reply:
Yes, I removed this.